### PR TITLE
JIT compiled csr/csc matvec and index mapping. 

### PR DESCRIPTION
**Pull request by weinbe58**
Adding jit compiled functions for inplace `y=aAx+y` for sparse matrices since these are not implemented in SciPy. 

The results are pretty much as good as we can probably get with pure python.

running this snippet of code as a benchmark:

```python
from bloqade import start
from cProfile import Profile
import numpy as np

pr = Profile()
pr.enable()


result = (
    start.add_position([(0,i*6.1) for i in range(14)]).rydberg.detuning.uniform.piecewise_linear(
        [0.1,3.8,0.1],[-20,-20,"d","d"]
    )
    .amplitude.uniform.piecewise_linear(
        [0.1,3.8,0.1],[0,15,15,0]
    )
    .phase.uniform.constant(np.pi/2, 4)
    .batch_assign(d=[1,2,4,8,16,32])
    .bloqade.python().run(100)
)

pr.disable()
pr.print_stats(sort="time")
```

on main this gives:

`         16206728 function calls (16142806 primitive calls) in 41.839 seconds`

after this PR 

`         7542139 function calls (7424975 primitive calls) in 30.869 seconds`

I think most of the optimization here is just cutting down on the number of function calls in python 😂

### CODE REVIEW

**Review by Roger-luo — APPROVED** [review]
I'm OK with the CSR operations added here, but we should warn somewhere in the docs or just module comments that the behavior and the corresponding module may get replaced with other things. 

At some point I'd like to refactor the emulation with Jax - I have something primitive for dynamics already, a more production-ready version may be more complicated than this PR due to the limitations of JAX.

We can probably just write our own CSR in JAX given this is already implemented in this PR. The reason is JAX can further offer auto-vectorize and AD through `diffrax`. Tho its sparse matrix and complex number solvers are quite primitive but could be something we contribute to the community.

File: src/bloqade/emulate/codegen/hamiltonian.py [hunk 1]
```diff
@@ -169,10 +172,14 @@ def matrix_ele(atom_index):
             indptr[1:] = indptr[:-1]
             indptr[0] = 0
 
-            operator = csr_matrix(
+            operator = SparseMatrixCSR.create(
```
Comment (Roger-luo) [inline src/bloqade/emulate/codegen/hamiltonian.py:175]: FYI. I think most people uses `new` for constructor (so it follows the same convention in C++, rust etc.), but prob not a big deal.

Comment (weinbe58) [reply src/bloqade/emulate/codegen/hamiltonian.py:175]: yeah this is an internal API so I don't think it matters so much.

**Comment by weinbe58** [comment]
> I'm OK with the CSR operations added here, but we should warn somewhere in the docs or just module comments that the behavior and the corresponding module may get replaced with other things.
> 
> At some point I'd like to refactor the emulation with Jax - I have something primitive for dynamics already, a more production-ready version may be more complicated than this PR due to the limitations of JAX.
> 
> We can probably just write our own CSR in JAX given this is already implemented in this PR. The reason is JAX can further offer auto-vectorize and AD through `diffrax`. Tho its sparse matrix and complex number solvers are quite primitive but could be something we contribute to the community.

OK lets try to work on this, I also think it would be super nice and @jon-wurtz can use it for his modeling efforts.

**Pull request merged**